File: day15_part2.py
#!/usr/bin/python3

import loader

inp = loader.string_list("input_day15")

number_list = [int(i) for i in inp[0].split(",")]

number_history = [None for i in range(30000000)]

for i in range(len(number_list)-1):
    number_history[number_list[i]] = i

last_said_number_index = len(number_list)-1
last_said_number = number_list[-1]

while True:
    previous_last_said_number_index = number_history[last_said_number]
    if previous_last_said_number_index == None:
        new_number = 0
    else:
        new_number = last_said_number_index - previous_last_said_number_index
    number_history[last_said_number] = last_said_number_index
    last_said_number = new_number
    last_said_number_index += 1
    if last_said_number_index == 30000000-1:
        print(last_said_number)
        break

File: day15_part2.c
#include <stdio.h>
#include <stdlib.h>

/*  A just-for-fun rewrite of day 15 part 2 in C. On my machine, the python
 *  solution takes ~8000ms, while this solution takes ~600ms with default gcc
 *  options and ~430ms with -O3.
 */

int main() {
  int numberList[] = {1,0,15,2,10,13}; // problem input
  int *numberHistory;
  int lastSaidNumberIndex = 5;
  int lastSaidNumber = numberList[5];
  
  numberHistory = (int *) malloc(30000000*sizeof(int));

  for (int i = 0; i <= 29999999; i++) {
    numberHistory[i] = -1;
  }

  for (int i = 0; i <= 5; i++) {
    numberHistory[numberList[i]] = i;
  }

  while(1) {
    int previousLastSaidNumberIndex = numberHistory[lastSaidNumber];
    int newNumber;
    
    if (previousLastSaidNumberIndex == -1) {
      newNumber = 0;
    } else {
      newNumber = lastSaidNumberIndex - previousLastSaidNumberIndex;
    }
    numberHistory[lastSaidNumber] = lastSaidNumberIndex;
    lastSaidNumber = newNumber;
    lastSaidNumberIndex++;
    if (lastSaidNumberIndex == (30000000-1)) {
      printf("Solution: %d",lastSaidNumber);
      break;
    }
  }
  return 0;
}
